Common/config/ruledtaskbarappearance.hpp

```diff
@@ -0,0 +1,117 @@
+#pragma once
+#include <string_view>
+#include <vector>
+#include <format>
+
+#include "rapidjsonhelper.hpp"
+#include "optionaltaskbarappearance.hpp"
+#include "rule.hpp"
+#include "../../TranslucentTB/windows/window.hpp"
+
+struct RuledTaskbarAppearance : OptionalTaskbarAppearance {
+	std::vector<Rule> Rules = {};
+
+	constexpr RuledTaskbarAppearance() noexcept = default;
+	constexpr RuledTaskbarAppearance(bool enabled, ACCENT_STATE accent, Util::Color color, bool showPeek, std::vector<Rule> rules) noexcept :
+		OptionalTaskbarAppearance(enabled, accent, color, showPeek),
+		Rules(rules)
+	{ }
+
+	template<typename Writer>
+	inline void Serialize(Writer& writer) const
+	{
+		OptionalTaskbarAppearance::Serialize(writer);
+		SerializeRulesVec(writer, Rules, RULES_KEY);
+	}
+
+	void Deserialize(const rjh::value_t& obj, void (*unknownKeyCallback)(std::wstring_view))
+	{
+		rjh::EnsureType(rj::Type::kObjectType, obj.GetType(), L"root node");
+
+		for (auto it = obj.MemberBegin(); it != obj.MemberEnd(); ++it)
+		{
+			rjh::EnsureType(rj::Type::kStringType, it->name.GetType(), L"member name");
+
+			const auto key = rjh::ValueToStringView(it->name);
+			if (key == RULES_KEY)
+			{
+				DeserializeRulesVec(it->value, Rules, key, unknownKeyCallback);
+			}
+			else
+			{
+				OptionalInnerDeserialize(key, it->value, unknownKeyCallback);
+			}
+		}
+	}
+
+	inline const std::optional<Rule> FindRule(Window window) const
+	{
+		for (const Rule rule : Rules)
+		{
+			// This is the fastest because we do the less string manipulation, so always try it first
+			if (!rule.m_Class.empty())
+			{
+				if (const auto className = window.classname())
+				{
+					if (rule.m_Class == *className)
+					{
+						return rule;
+					}
+				}
+			}
+
+			if (!rule.m_ProcessName.empty())
+			{
+				if (const auto file = window.file())
+				{
+					if (rule.m_ProcessName == file->filename().native())
+					{
+						return rule;
+					}
+				}
+			}
```

Comment:
it compiles fine now, but I don't know why the rules are not working correctly. 

I have this rule in the `maximized_window_appearance` section
```json
{
  "process_name": "Code.exe",
  "accent": "opaque",
  "color": "#1f1f1f"
}
```
and when I logged `rule.m_ProcessName` and `file->filename().native()`, I get `Code.exe` and `explorer.exe` respectively, any idea why TTB picks up `explorer.ex` and not `Code.exe`? 

Also does the `maximized_window_appearance` configuration gets applied each time the active window changes or not ?

Reply:
It's because you're passing the taskbar window's handle, not the handle of the current foreground window:

https://github.com/TranslucentTB/TranslucentTB/blob/93dfb2c563fef4e00f5839550a28f1e9f7e94c2c/TranslucentTB/taskbar/taskbarattributeworker.cpp#L229

https://github.com/TranslucentTB/TranslucentTB/blob/93dfb2c563fef4e00f5839550a28f1e9f7e94c2c/TranslucentTB/taskbar/taskbarattributeworker.cpp#L434

> Also does the maximized_window_appearance configuration gets applied each time the active window changes or not ?

No, currently we do not do that because there was no reason to track the foreground window. For now, you can ignore that bit and I'll add it, I already have code doing that somewhere.

Reply:
> It's because you're passing the taskbar window's handle, not the handle of the current foreground window:

I don't know what I was thinking, it literally says `TaskbarWindow` LOL

> > Also does the maximized_window_appearance configuration gets applied each time the active window changes or not ?
> 
> No, currently we do not do that because there was no reason to track the foreground window. For now, you can ignore that bit and I'll add it, I already have code doing that somewhere.

Without triggering `GetConfig`  when foreground window change, I am having a hard time testing if this PR works or not tbh. Anyways I fixed the serialization and passed correct window.

Reply:
I restored the foreground window tracking code, you should be able to use it.

Reply:
thanks